### PR TITLE
 Rename "builtin versions" to "builtin semantics variants" to reduce risk of confusion

### PR DESCRIPTION
**Pull request by kwxm**
[PLT-7306]

The `ToBuiltinMeaning` class in PlutusCore.Builtin.Meaning has an associated class called `BuiltinVersion` and in PlutusCore.Default.Builtins this is instantiated as `data BuiltinVersion DefaultFun = DefaultFunV1 | DefaultFunV2`
This risks confusion with PlutusV1, PlutusV2 etc. (as seen in [this issue](https://github.com/input-output-hk/plutus/issues/5475)).

After some Slack discussion we’ve decided to rename `BuiltinVersion` to `BuiltinSemanticsVariant` and `DefaultFunV1` to `DefaultFunSemanticsVariant1` , etc. I've also tried to remove all mentions of "builtin versions" in comments.

[PLT-7306]: https://input-output.atlassian.net/browse/PLT-7306?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: plutus-ledger-api/src/PlutusLedgerApi/Common/Eval.hs [hunk 1]
```diff
@@ -119,17 +119,21 @@ newtype EvaluationContext = EvaluationContext
     deriving stock Generic
     deriving anyclass (NFData, NoThunks)
 
-{-|  Create an 'EvaluationContext' for a given plutus-builtins' version.
+{-|  Create an 'EvaluationContext' for a given builtin semantics variant.
```
Comment (kwxm) [inline plutus-ledger-api/src/PlutusLedgerApi/Common/Eval.hs:123]: Is that right?  I presume it's not referring to the ledger language version, although this is in fact what gets called in the V1/V2/V2 `mkEvaluationContext` functions.

File: plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs [hunk 2]
```diff
@@ -70,57 +70,76 @@ data BuiltinMeaning val cost =
...
+typeOfBuiltinFunction semvar fun =
+    case toBuiltinMeaning @_ @_ @(Term TyName Name uni fun ()) semvar fun of
         BuiltinMeaning sch _ _ -> typeSchemeToType sch
 
-{- Note [Versioned builtins]
-The purpose of the "versioned builtins" feature is to provide multiple, different denotations
-(implementations) for the same builtin(s).
-An example use of this feature is for "fixing" the behaviour of `ConsByteString` builtin to throw an
-error instead of overflowing its first argument.
+{- Note [Builtin semantics variants] The purpose of the "builtin semantics
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:106]:
```suggestion
{- Note [Builtin semantics variants] 
The purpose of the "builtin semantics
```

File: plutus-ledger-api/src/PlutusLedgerApi/Common/Eval.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-ledger-api/src/PlutusLedgerApi/Common/Eval.hs:123]: Yes, I think it is right. It's getting called with the corresponding `BuiltinSemanticsVariant`. Possibly we should just use this generically, and instead have for each language a clearly associated `BuiltinSemanticsVariant` or something.

**Review by michaelpj — APPROVED** [review]
Tedious, but good!

File: plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs (hunk 2 above)
Comment (kwxm) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs]: Oh, that seems to have broken something.  I'll try to fix it manually.

**Pull request merged**